### PR TITLE
added authentication

### PR DESCRIPTION
**Pull request by ahijaouy**
Implemented Passport authentication to protect routes.

### CODE REVIEW

File: backend/routes/users.js [hunk 1]
```diff
@@ -2,41 +2,70 @@
...
+      res.status(500).send(errors);
+    });
+});
+
+//************** LOGIN WALL *******************
+router.use((req, res, next) => {
+  if (req.user) {
+    next();
+  } else {
+    res.status(401).send('YOU MUST BE AUTHENTICATED TO ACCESS THIS ROUTE');
```
Comment (jvt) [inline backend/routes/users.js]: return callbacks. aggressive error messages++

File: backend/routes/users.js [hunk 2]
```diff
@@ -2,41 +2,70 @@
...
+  if (!errors.isEmpty()) {
+    return res.status(400).json({ errors: errors.mapped() });
+  }
+  const userData = matchedData(req);
+  let emailInUse = false;
+  User.findOne({ email: userData.email })
+    .then(user => {
+      if (user) {
+        emailInUse = true;
+        throw new Error("Email already in use")
```
Comment (jvt) [inline backend/routes/users.js]: Feels somewhat janky but I'm okay with it

File: backend/routes/index.js [hunk 3]
```diff
@@ -27,83 +27,75 @@ router.use(session({
...
+
+// Login Failed Route
+router.get('/failure', (req, res) => {
+  res.status(400).json({ authenticated: false });
+});
+
+// Logout Route
+router.get('/logout', (req, res) => {
+  req.logout();
+  res.json({ logout: 'success' });
```
Comment (jvt) [inline backend/routes/index.js]: return yo callbacks

File: backend/routes/index.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ const emails = require('./emails');
 const responses = require('./responses');
 const surveys = require('./surveys');
 const users = require('./users');
-
+``
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/index.js]: lol oops

File: backend/routes/index.js [hunk 5]
```diff
@@ -27,83 +27,75 @@ router.use(session({
...
+      if (err) return done(err);
+      if (!user || !user.verifyPassword(password)) {
+        return done(null, false, { message: 'Login Error.' });
+      }
+      return done(null, user);
+    });
+  }));
+
+passport.serializeUser((user, done) => {
+  done(null, user.id);
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/index.js]: One day I'll get that right on the first try...

File: backend/routes/index.js [hunk 6]
```diff
@@ -27,83 +27,75 @@ router.use(session({
...
+});
+
+passport.deserializeUser((id, done) => {
+  User.findById(id, (err, user) => {
+    done(err, user);
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/index.js]: done

File: backend/routes/index.js [hunk 7]
```diff
@@ -27,83 +27,75 @@ router.use(session({
...
+  });
+});
+
+
+// Login Route
+router.post('/login',
+  passport.authenticate('local', {
+    successRedirect: '/api/success',
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/index.js]: Yes they do. This entire routes/index.js exists at the ```/api/``` endpoint. THat's specified in the ```server.js```

File: backend/routes/index.js [hunk 8]
```diff
@@ -27,83 +27,75 @@ router.use(session({
...
+    failureRedirect: '/api/failure'
+  })
+);
+
+// Login Failed Route
+router.get('/failure', (req, res) => {
+  res.status(400).json({ authenticated: false });
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/index.js]: I like the way you suggested! I switched it!

File: backend/routes/index.js (hunk 3 above)
Comment (ahijaouy) [reply backend/routes/index.js]: done

File: backend/routes/index.js [hunk 9]
```diff
@@ -27,83 +27,75 @@ router.use(session({
...
-// );
-
-// // Login Failed Route
-// router.get('/failed', (req, res) => {
-//     res.status(400).json({authenticated: false});
-// });
+//************** LOGIN WALL *******************
+router.use((req, res, next) => {
+  if (req.user) {
+    next();
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/index.js]: done

File: backend/routes/index.js [hunk 10]
```diff
@@ -27,83 +27,75 @@ router.use(session({
...
+  } else {
+    res.status(401).send('YOU MUST BE AUTHENTICATED TO ACCESS THIS ROUTE');
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/index.js]: Done. And lol meh ;P

File: backend/routes/index.js [hunk 11]
```diff
@@ -27,83 +27,75 @@ router.use(session({
...
+});
 
-// // Logout Route
-// router.get('/logout', (req, res) => {
-//     req.logout();
-//     res.json({ logout: 'success' });
-// });
+// Login Success Route
+router.get('/success', (req, res) => {
+  res.json({ authenticated: true, user: req.user });
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/index.js]: fixed it for you bb

File: backend/routes/users.js (hunk 2 above)
Comment (ahijaouy) [reply backend/routes/users.js]: I wasn't sure how to tell from the Error message in the catch block what caused the error. I only wanna tell them that their email is in use but no other error message info... that's how I thought of doing it.. Suggestions?

File: backend/routes/users.js [hunk 12]
```diff
@@ -2,41 +2,70 @@
...
+        return res.status(400).json({errors: "Email in use"});
+      }
+      res.status(500).send(errors);
+    });
+});
+
+//************** LOGIN WALL *******************
+router.use((req, res, next) => {
+  if (req.user) {
+    next();
```
[the remark this answers is not in the document]
Comment (ahijaouy) [reply backend/routes/users.js]: fixed

File: backend/routes/users.js (hunk 1 above)
Comment (ahijaouy) [reply backend/routes/users.js]: fixed

**Pull request merged**